### ISSUE TITLE
Implement new schema updates to tagbase-schema.sql to accomodate multiple position dataset solutions per tag deployment

### ISSUE THREAD

**Issue by lewismc** [tagbase/tagbase-server#99]
From Vardis' input, the proposed changes to [tagbase-schema.sql](https://github.com/tagbase/tagbase-server/blob/main/services/postgres/sqldb/tagbase-schema.sql) are motivated by

* The scenario of multiple position dataset solutions per tag deployment
* ARGOS location class codes reflecting lat/lon position record accuracies that we were not handling previously and that resulted in the ingest error Lewis reported in Slack and that Camryn had mentioned previously.

The updates are illustrated in the attached Tagbase entity-relation diagram below. But here is a brief summary of what Vardis proposed.

* **proc_observations table**: changed the “_variable_value_” field type from _double_ precision to _varchar(30)_ which will now allow for the ARGOS alphanumeric location code (_A,B,C,Z,0,1,2,3_ per page 13 table <a href="https://www.argos-system.org/wp-content/uploads/2016/08/r363_9_argos_users_manual-v1.6.6.pdf" style="color:rgb(5,99,193)" target="_blank" data-saferedirecturl="https://www.google.com/url?q=https://www.argos-system.org/wp-content/uploads/2016/08/r363_9_argos_users_manual-v1.6.6.pdf&amp;source=gmail&amp;ust=1652046159480000&amp;usg=AOvVaw0wbIDN52EbgIJKrIR2SA_Z">https://www.argos-system.org/<wbr>wp-content/uploads/2016/08/<wbr>r363_9_argos_users_manual-v1.<wbr>6.6.pdf</a>) to be ingested in this temporary staging table of eTUFF data.  (can make field &gt;30 characters if higher precision would be desirable). The implication of this change is that some type checking should occur when data are moved from this temporary “proc_” staging table to the target destination “data_” tables.
* **observation_types** table: I added a record with _variable_id=0_ that is a new entry for the ARGOS location class variable type as follows (so we will ultimately need to update this table listing also in Github).

| variable_id    |     variable_name         |     standard_name    |     variable_source    |     variable_units    |     notes                                                                                                                                                      |     standard_unit |
|----------------|---------------------------|----------------------|------------------------|-----------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------|
|           0    |     ArgosLocationClass    |                      |     common             |     unitless          |     ARGOS Location Class code (G,3,2,1,0,A,B,Z)    https://www.argos-system.org/wp-content/uploads/2016/08/r363_9_argos_users_manual-v1.6.6.pdf  , page 13.    |                   |

* **data_position table**: the following updates were made:
    * Added a new “_argos_location_class_” field of _type varchar(1)_ that will take the argos position quality code for each lat/lon record for a given unique _submission_, _tag_id_ and _geolocation solution dataset_ if those quality codes are
      available (entries into this column is not a required).
    * Added a “_solution_id_” field that will for a given _submission_id_ and _tag_id_ will differentiate between potentially multiple associated positional dataset solutions.  Note that as a convention, a _solution_id=1_  will be assigned to the primary/”best” set of positional data for a given tag dataset.  So when we want a materialized view with the accepted best positions and tag measurement data then a filter will be applied to the query to just select positional records for the tag with _solution_id=1_. Note that by convention, we should have the primary positional solution and tag measurement data and metadata submitted together as a single, consolidated eTUFF file.  If there are additional positional solution datasets, then they each should be submitted as standalone eTUFF files with just the _lat_, _lon_ , _time_, _lat_error_, _lon_error_, _argos_location_class data and associated key metadata (tag identifier attributes and waypoint metadata).  The solution_id would be incremented for each of these supplemented positional solution datasets for a given tag_id and submission_id.
* **submission table**: I also removed the “_dmas_granule_id_” field from the which is no longer relevant as we discussed last time. To that table I did also add a new “_notes_” text field for users potentially to include any free-form descriptive information on the ingest if useful.

Might take a moment to digest all this, but I think this is probably the cleanest and most parsimonious way to handle multiple positional datasets per tag deployment with relatively minimal changes to the schema. This should also allow for easy extraction of the primary positional solution or any other solutions for a given tag for possible comparison easily.

**Comment by lewismc** [comment]
![TagbaseERD_20220403](https://user-images.githubusercontent.com/1165719/167272953-b26e09e2-9506-41c1-87d7-2fdd85254c62.gif)

**Comment by lewismc** [comment]
Further input from Vardis,

----------
I took a look at the revised tagbase schema from a couple of weeks ago (described in the initial email of this thread) and just made the following 2 changes:
- **data_position table**: added the "_solution_id_" field to the primary key (so now _submission_id_, _tag_id_, _solution_id_, and _date_time_ all uniquely identify a row with positional for a given tag and track solution submitted)
- given this modified primary key constraint, foreign keys in the related tag data tables (data_profiles, data_time_series, data_histogram_bin_data) needed to be dropped because uniqueness referential integrity constraints no longer were fulfilled.  This is not an issue.

Attached below is the updated Entity Relation Model or schema diagram (.gif)

Note that the structure of the "_metadata_position_" table in this output is in the normalized form I had initially proposed, and not in the denormalized table form Tim's subsequently suggested. Tim, if you could send me the link to the GoogleSheet where you show how this table is structured then I can try to implement a separate version of the schema with that approach instead for Lewis to then decide which is best also from an implementation perspective.  We can achieve the same goal of filtering out position data for a given tag that is considered/flagged in the metadata_position table structured in either way.

One other thing Tim, I left the "ArgosLocationClass" record in the "observation_types" table with a "variable_id" of 0 for the time being.  Note that the number of records in that table are somewhat different from the csv listing on your Tagbase githum repo (https://github.com/tagbase/tagbase/blob/master/etuff/eTUFF-ObservationTypes.csv).  We should double check this for consistency.  Then I can update the ArgosLocationClass variable id if necessary - using a 0 is fine and just as legitimate as any other permissible integer.

**Comment by lewismc** [comment]
Comment from Tim

----------------
Thanks Vardis for doing these! They are all very clear to me, and I am OK with the changes.

The only thing I will raise is the numbering of "_ArgosLocationClass_". I would prefer to stick to the convention that we told Cameron a while back if he were to make new additions, we'd like it to increment from the "last" ID number, and that's what he did for a few items he added, and subsequently, I have incorporated his edits, and everything is now synched up with his version (i.e., the github link you have listed). Using "0" is ok - but then it's a deviation. Breaking from the convention, to me, is something to be done when something is very important, but LocationClass isn't.

**Comment by lewismc** [comment]
Comment from Vardis
------------
> Just as I mentioned, there is a bit of a difference in the number of records currently in that observation_types table relative to the csv on your Tagbase GitHub.  So we should check that, and then can update that Argos code accordingly.

**Comment by lewismc** [comment]
Post #100 the current ingestion errors are as follows
```
postgres_1        | 2022-05-08 22:48:10.118 UTC [79] ERROR:  duplicate key value violates unique constraint "observation_types_pkey"
postgres_1        | 2022-05-08 22:48:10.118 UTC [79] DETAIL:  Key (variable_id)=(2) already exists.
postgres_1        | 2022-05-08 22:48:10.118 UTC [79] STATEMENT:  INSERT INTO observation_types(variable_name, variable_units) VALUES ('argosLC', 'unitless') ON CONFLICT (variable_name) DO NOTHING
tagbase_server_1  | ERROR:__main__:Exception on /v0.5.0/ingest [GET]
tagbase_server_1  | Traceback (most recent call last):
tagbase_server_1  |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 2077, in wsgi_app
tagbase_server_1  |     response = self.full_dispatch_request()
tagbase_server_1  |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 1525, in full_dispatch_request
tagbase_server_1  |     rv = self.handle_user_exception(e)
tagbase_server_1  |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 1523, in full_dispatch_request
tagbase_server_1  |     rv = self.dispatch_request()
tagbase_server_1  |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 1509, in dispatch_request
tagbase_server_1  |     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
tagbase_server_1  |   File "/usr/local/lib/python3.10/site-packages/connexion/decorators/decorator.py", line 68, in wrapper
tagbase_server_1  |     response = function(request)
tagbase_server_1  |   File "/usr/local/lib/python3.10/site-packages/connexion/decorators/uri_parsing.py", line 149, in wrapper
tagbase_server_1  |     response = function(request)
tagbase_server_1  |   File "/usr/local/lib/python3.10/site-packages/connexion/decorators/validation.py", line 399, in wrapper
tagbase_server_1  |     return function(request)
tagbase_server_1  |   File "/usr/local/lib/python3.10/site-packages/connexion/decorators/parameter.py", line 120, in wrapper
tagbase_server_1  |     return function(**kwargs)
tagbase_server_1  |   File "/usr/src/app/tagbase_server/controllers/ingest_controller.py", line 127, in ingest_get
tagbase_server_1  |     cur.execute(
tagbase_server_1  | psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "observation_types_pkey"
tagbase_server_1  | DETAIL:  Key (variable_id)=(2) already exists.
tagbase_server_1  |
tagbase_server_1  | ERROR:tornado.access:500 GET /v0.5.0/ingest?file=file%3A%2F%2F%2Fusr%2Fsrc%2Fapp%2Fdata%2F159924_2013_128419_eTUFF_hdr.txt&type=etuff (192.168.80.1) 60.83ms
```
I am going to keep this issue open until I have fixed this for each eTUFF file I have at my disposal. Testing is ongoing.

**Comment by lewismc** [comment]
The new `version` (defaults to 1) and `notes` columns were implemented in the REST API in #101
